File: batou-rules/testdata/fixtures/php/vulnerable/log_injection.php
<?php
// Log injection via unsanitized user input in logging functions.
// Vulnerable: user input can inject newlines to forge log entries.

$username = $_POST['username'];
$action = $_GET['action'];

// GTSS-LOG-001: User input in error_log without sanitization
error_log("User login attempt: " . $username . " from " . $_SERVER['REMOTE_ADDR']);

// GTSS-LOG-001: User input in syslog
syslog(LOG_INFO, "Action performed: " . $action . " by user " . $username);

// GTSS-LOG-003: Logging sensitive data - password in log
$password = $_POST['password'];
error_log("Login attempt with password: " . $password);

// Laravel-style logging with user input
// Log::info("User search: " . $_GET['query']);

// CRLF injection: user can inject newlines to forge log entries
$search_term = $_GET['q'];
error_log("[SEARCH] User searched for: " . $search_term);

// Sensitive data logged
$api_key = $_POST['api_key'];
error_log("API request with api_key: " . $api_key);

// Combined: user input + sensitive data in syslog
$token = $_COOKIE['auth_token'];
syslog(LOG_WARNING, "Token validation for token: " . $token . " user: " . $_GET['user']);

File: batou-rules/testdata/fixtures/csharp/vulnerable/LogInjection.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;

namespace Vulnerable.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class LogController : ControllerBase
    {
        private readonly ILogger<LogController> _logger;

        public LogController(ILogger<LogController> logger)
        {
            _logger = logger;
        }

        [HttpPost("login")]
        public IActionResult Login(string username)
        {
            // Vulnerable: user input directly concatenated into log message
            _logger.LogInformation("Login attempt for user: " + username);
            _logger.LogWarning("Failed login for: " + username);
            return Ok();
        }
    }
}
